src/solidity/typeDecoder.js

```diff
@@ -0,0 +1,288 @@
+'use strict'
+
+/**
+  * determine if the given @arg type need to start from a new slot
+  *
+  * @param {Object} type - current type (object returned by the getType function)
+  * @return true if the type needs a new slot
+  */
+function typeNeedNewSlot (type) {
+  return (type.storageBytes === 'dynamic' || type.arraySize || type.members || type.fullType === 'string' || type.fullType === 'bytes')
+}
+
+/**
+  * parse the full type
+  *
+  * @param {String} fullType - type given by the AST (ex: uint[2] storage ref[2])
+  * @return {String} return the first part of the full type. keep the array declaration (uint[2][2] in that case)
+  */
+function extractType (fullType) {
+  var dim = []
+  if (fullType.indexOf('[') !== -1) {
+    dim = getTotalDimensions(fullType)
+  }
+  if (fullType.indexOf(' ') !== -1) {
+    var splitted = fullType = fullType.split(' ')
+    if (fullType.indexOf('struct') === 0 || fullType.indexOf('enum') === 0) {
+      fullType = splitted[0] + ' ' + splitted[1]
+    } else {
+      fullType = splitted[0]
```

Comment:
Anything that is not a struct or enum is ignored? I think we should _really_ add some proper error reporting.

We could turn this whole thing into a class and report the variables and contracts we were not able to handle properly.